server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1813,17 +1825,19 @@ record RootBlobUpdateResult(RepositoryData oldRepositoryData, RepositoryData new
 
                     // Write global metadata
                     final Metadata clusterMetadata = finalizeSnapshotContext.clusterMetadata();
-                    executor.execute(
-                        ActionRunnable.run(
-                            allMetaListeners.acquire(),
-                            () -> GLOBAL_METADATA_FORMAT.write(clusterMetadata, blobContainer(), snapshotId.getUUID(), compress)
-                        )
-                    );
+                    final var projectMetadata = clusterMetadata.getProject(getProjectId());
+                    executor.execute(ActionRunnable.run(allMetaListeners.acquire(), () -> {
+                        if (finalizeSnapshotContext.serializeProjectMetadata()) {
+                            PROJECT_METADATA_FORMAT.write(projectMetadata, blobContainer(), snapshotId.getUUID(), compress);
+                        } else {
+                            GLOBAL_METADATA_FORMAT.write(clusterMetadata, blobContainer(), snapshotId.getUUID(), compress);
+                        }
+                    }));
```

Comment:
This is where we conditionally write `ProjectMetadata` for multi-project snapshots. The `Metadata` in this case is a thin wrapper around `ProjectMetadata` to reuse existing finalization related classes.